Index: linux-2.6.21.5/drivers/i2c/chips/pcf50606.c
===================================================================
--- linux-2.6.21.5.orig/drivers/i2c/chips/pcf50606.c	2007-07-16 12:21:35.724785054 +0200
+++ linux-2.6.21.5/drivers/i2c/chips/pcf50606.c	2007-07-16 12:21:57.510026525 +0200
@@ -1871,12 +1871,12 @@
 static struct i2c_driver pcf50606_driver = {
 	.driver = {
 		.name	= "pcf50606",
-		.suspend= &pcf50606_suspend,
-		.resume	= &pcf50606_resume,
+		.suspend= pcf50606_suspend,
+		.resume	= pcf50606_resume,
 	},
 	.id		= I2C_DRIVERID_PCF50606,
-	.attach_adapter	= &pcf50606_attach_adapter,
-	.detach_client	= &pcf50606_detach_client,
+	.attach_adapter	= pcf50606_attach_adapter,
+	.detach_client	= pcf50606_detach_client,
 };
 
 /* platform driver, since i2c devices don't have platform_data */
