Index: linux-2.6.17.14-fic4.test/drivers/input/touchscreen/s3c2410_ts.c
===================================================================
--- linux-2.6.17.14-fic4.test.orig/drivers/input/touchscreen/s3c2410_ts.c	2007-01-15 01:07:12.000000000 +0100
+++ linux-2.6.17.14-fic4.test/drivers/input/touchscreen/s3c2410_ts.c	2007-01-15 01:07:44.000000000 +0100
@@ -190,7 +190,8 @@
 		writel(S3C2410_ADCTSC_PULL_UP_DISABLE | AUTOPST, base_addr+S3C2410_ADCTSC);
 		writel(readl(base_addr+S3C2410_ADCCON) | S3C2410_ADCCON_ENABLE_START, base_addr+S3C2410_ADCCON);
 	} else {
-		mod_timer(&touch_timer, jiffies+1);
+		//mod_timer(&touch_timer, jiffies+1);
+		mod_timer(&touch_timer, jiffies+(HZ/2));
 		writel(WAIT4INT(1), base_addr+S3C2410_ADCTSC);
 	}
 
