Index: omoko/src/dates_omoko.c
===================================================================
--- omoko.orig/src/dates_omoko.c	2007-01-26 13:48:10.000000000 +0100
+++ omoko/src/dates_omoko.c	2007-01-26 13:58:35.000000000 +0100
@@ -28,7 +28,6 @@
 #include <libecal/e-cal.h>
 #include <libecal/e-cal-time-util.h>
 #include <libical/ical.h>
-#include <libedataserver/e-data-server-util.h>
 
 #include <string.h>
 
