Index: linux-2.6.22.1/arch/arm/mach-s3c2410/mach-gta01.c
===================================================================
--- linux-2.6.22.1.orig/arch/arm/mach-s3c2410/mach-gta01.c
+++ linux-2.6.22.1/arch/arm/mach-s3c2410/mach-gta01.c
@@ -486,8 +486,8 @@
 
 static struct s3c2410_ts_mach_info gta01_ts_cfg = {
 	.delay = 10000,
-	.presc = 49,
-	.oversampling_shift = 2,
+	.presc = 65,
+	.oversampling_shift = 5,
 };
 
 /* SPI */
