Index: linux-2.6.20/drivers/base/power/main.c
===================================================================
--- linux-2.6.20.orig/drivers/base/power/main.c	2007-02-15 19:56:34.000000000 +0100
+++ linux-2.6.20/drivers/base/power/main.c	2007-02-15 19:56:43.000000000 +0100
@@ -53,8 +53,8 @@
 {
 	int error;
 
-	pr_debug("PM: Adding info for %s:%s\n",
-		 dev->bus ? dev->bus->name : "No Bus", dev->kobj.name);
+	//pr_debug("PM: Adding info for %s:%s\n",
+	//	 dev->bus ? dev->bus->name : "No Bus", dev->kobj.name);
 	down(&dpm_list_sem);
 	list_add_tail(&dev->power.entry, &dpm_active);
 	device_pm_set_parent(dev, dev->parent);
