diff -Nru --exclude-from=/sunbeam/home/laforge/scripts/dontdiff linux-2.6.17.7-new/drivers/mmc/s3c2410mci.c linux-2.6.17.7-qt/drivers/mmc/s3c2410mci.c
--- linux-2.6.17.7-new/drivers/mmc/s3c2410mci.c	2006-08-07 09:25:32.000000000 +0200
+++ linux-2.6.17.7-qt/drivers/mmc/s3c2410mci.c	2006-08-07 00:55:23.000000000 +0200
@@ -206,6 +206,8 @@
 	host->complete_what = COMPLETION_NONE;
 	complete(&host->complete_request);
 	writel(0, host->base + S3C2410_SDIIMSK);
+	if (host->mrq->data && host->mrq->data->error != MMC_ERR_NONE)
+		complete(&host->complete_dma);
 	spin_unlock_irqrestore( &host->complete_lock, iflags);
 	DBG(PFX "IRQ transfer closed.\n");
 	return IRQ_HANDLED;

