Enlarge junction circles in Postscript "Plot" of schematics.

From: Alvaro Lopes <alvieboy@alvie.com>

---

--- kicad.orig/eeschema/plotps.cpp.orig	2009-06-18 09:04:51.779811920 +0100
+++ kicad/eeschema/plotps.cpp	2009-06-18 09:05:24.726650982 +0100
@@ -515,7 +515,7 @@
                 #define STRUCT ( (DrawJunctionStruct*) DrawList )
             if( g_PlotPSColorOpt )
                 SetColorMapPS( ReturnLayerColor( STRUCT->GetLayer() ) );
-            PlotCercle( STRUCT->m_Pos, DRAWJUNCTION_SIZE, 1 );
+            PlotCercle( STRUCT->m_Pos, DRAWJUNCTION_SIZE*2, 1 );
             break;
 
         case TYPE_SCH_TEXT:
