Index: linux-2.6.17.14/sound/soc/s3c24xx/s3c24xx-i2s.c
===================================================================
--- linux-2.6.17.14.orig/sound/soc/s3c24xx/s3c24xx-i2s.c
+++ linux-2.6.17.14/sound/soc/s3c24xx/s3c24xx-i2s.c
@@ -60,13 +60,13 @@ static struct s3c2410_dma_client s3c24xx
 
 static s3c24xx_pcm_dma_params_t s3c24xx_i2s_pcm_stereo_out = {
 	.client		= &s3c24xx_dma_client_out,
-	.channel	= DMACH_I2S_OUT,
+	.channel	= 2,
 	.dma_addr	= S3C2410_PA_IIS + S3C2410_IISFIFO
 };
 
 static s3c24xx_pcm_dma_params_t s3c24xx_i2s_pcm_stereo_in = {
 	.client		= &s3c24xx_dma_client_in,
-	.channel	= DMACH_I2S_IN,
+	.channel	= 1,
 	.dma_addr	= S3C2410_PA_IIS + S3C2410_IISFIFO
 };
 
Index: linux-2.6.17.14/sound/soc/s3c24xx/s3c24xx-pcm.c
===================================================================
--- linux-2.6.17.14.orig/sound/soc/s3c24xx/s3c24xx-pcm.c
+++ linux-2.6.17.14/sound/soc/s3c24xx/s3c24xx-pcm.c
@@ -119,9 +119,9 @@ static void s3c24xx_pcm_enqueue(struct s
 	prtd->dma_pos = pos;
 }
 
-void s3c24xx_audio_buffdone(struct s3c2410_dma_chan *channel,
+void s3c24xx_audio_buffdone(s3c2410_dma_chan_t *channel,
 							void *dev_id, int size,
-							enum s3c2410_dma_buffresult result)
+							s3c2410_dma_buffresult_t result)
 {
 	struct snd_pcm_substream *substream = dev_id;
 	struct s3c24xx_runtime_data *prtd = substream->runtime->private_data;
@@ -260,7 +260,7 @@ static int s3c24xx_pcm_trigger(struct sn
 	case SNDRV_PCM_TRIGGER_PAUSE_RELEASE:
 		prtd->state |= ST_RUNNING;
 		s3c2410_dma_ctrl(prtd->params->channel, S3C2410_DMAOP_START);
-		s3c2410_dma_ctrl(prtd->params->channel, S3C2410_DMAOP_STARTED);
+//		s3c2410_dma_ctrl(prtd->params->channel, S3C2410_DMAOP_STARTED);
 		break;
 
 	case SNDRV_PCM_TRIGGER_STOP:
Index: linux-2.6.17.14/sound/soc/s3c24xx/Kconfig
===================================================================
--- linux-2.6.17.14.orig/sound/soc/s3c24xx/Kconfig
+++ linux-2.6.17.14/sound/soc/s3c24xx/Kconfig
@@ -21,7 +21,7 @@ config SND_S3C24XX_SOC_SMDK2440
 
 config SND_S3C24XX_SOC_SMDK2440_WM8753
 	tristate "SoC I2S Audio support for S3C24XX - WM8753"
-	depends on SND_S3C24XX_SOC && MACH_SMDK
+	depends on SND_S3C24XX_SOC && MACH_GTA01
 	select SND_S3C24XX_SOC_I2S
 	help
 	  Say Y if you want to add support for SoC audio on smdk2440
