Index: linux-2.6.17.14-fic4.test/arch/arm/mach-s3c2410/mach-gta01.c
===================================================================
--- linux-2.6.17.14-fic4.test.orig/arch/arm/mach-s3c2410/mach-gta01.c	2007-01-08 22:47:48.000000000 +0100
+++ linux-2.6.17.14-fic4.test/arch/arm/mach-s3c2410/mach-gta01.c	2007-01-08 22:51:22.000000000 +0100
@@ -57,6 +57,7 @@
 #include <asm/arch/ts.h>
 #include <asm/arch/spi.h>
 #include <asm/arch/spi-gpio.h>
+#include <asm/arch/mci.h>
 
 #include <asm/arch/gta01.h>
 
@@ -232,7 +233,6 @@
 	.sets		= gta01_nand_sets,
 };
 
-#if 0
 static unsigned int mmc_millivolts[] = {
 	[MMC_VDD_145_150]	= 1500,
 	[MMC_VDD_150_155]	= 1500,
@@ -257,14 +257,15 @@
 	[MMC_VDD_32_33]		= 3300,
 };
 
-static void gta01_mmc_set_power(unsigned int to)
+static void gta01_mmc_set_power(unsigned short vdd)
 {
 	/* FIXME: configure D2REG of PCF50606 using D2REGC1 */
-	//pcf50606_voltage_set(mmc_millivolts[to]);
+	/* FIXME: set GTA01_GPIO_SDMMC_ON on GTA01v4 */
+	//pcf50606_voltage_set(mmc_millivolts[vdd]);
 }
 
-static struct s3c24xx_mmc_platdata gta01_mmc_cfg = {
-	.gpio_detect	= S3C2410_GPF5,
+static struct s3c24xx_mci_pdata gta01_mmc_cfg = {
+	.gpio_detect	= GTA01_GPIO_nSD_DETECT,
 	.set_power	= &gta01_mmc_set_power,
 	.ocr_avail	= MMC_VDD_145_150|MMC_VDD_150_155|MMC_VDD_155_160|
 			  MMC_VDD_160_165| MMC_VDD_165_170|MMC_VDD_17_18|
@@ -274,7 +275,6 @@
 			  MMC_VDD_27_28|MMC_VDD_28_29|MMC_VDD_29_30|
 			  MMC_VDD_30_31|MMC_VDD_31_32|MMC_VDD_32_33,
 };
-#endif
 
 static void gta01_udc_pullup(enum s3c2410_udc_cmd_e cmd)
 {
@@ -441,7 +441,7 @@
 static void __init gta01_machine_init(void)
 {
 	s3c_device_nand.dev.platform_data = &gta01_nand_info;
-	//s3c_device_sdi.dev.platform_data = &gta01_mmc_cfg;
+	s3c_device_sdi.dev.platform_data = &gta01_mmc_cfg;
 
 	s3c24xx_fb_set_platdata(&gta01_lcd_cfg);
 
