Index: xserver/configure.ac
===================================================================
--- xserver.orig/configure.ac	2007-11-23 17:32:58.000000000 +0800
+++ xserver/configure.ac	2007-11-23 17:33:11.000000000 +0800
@@ -1801,6 +1801,7 @@
 hw/kdrive/epson/Makefile
 hw/kdrive/fake/Makefile
 hw/kdrive/fbdev/Makefile
+hw/kdrive/glamo/Makefile
 hw/kdrive/i810/Makefile
 hw/kdrive/linux/Makefile
 hw/kdrive/mach64/Makefile
Index: xserver/hw/kdrive/Makefile.am
===================================================================
--- xserver.orig/hw/kdrive/Makefile.am	2007-11-23 17:32:58.000000000 +0800
+++ xserver/hw/kdrive/Makefile.am	2007-11-23 17:33:01.000000000 +0800
@@ -4,7 +4,7 @@
 endif
 
 if KDRIVEFBDEV
-FBDEV_SUBDIRS = fbdev
+FBDEV_SUBDIRS = fbdev glamo
 endif
 
 if XSDLSERVER
@@ -29,4 +29,4 @@
 	fake
 
 DIST_SUBDIRS = vesa ati chips epson i810 mach64 mga neomagic nvidia pm2 r128 \
-               smi via fbdev sdl ephyr src linux fake sis300
+               smi via fbdev sdl ephyr src linux fake sis300 glamo
