Index: linux-2.6.17.14-fic4.test/drivers/usb/gadget/ether.c
===================================================================
--- linux-2.6.17.14-fic4.test.orig/drivers/usb/gadget/ether.c	2007-01-21 21:31:48.000000000 +0100
+++ linux-2.6.17.14-fic4.test/drivers/usb/gadget/ether.c	2007-01-21 21:34:31.000000000 +0100
@@ -462,7 +462,7 @@
 	.bConfigurationValue =	DEV_CONFIG_VALUE,
 	.iConfiguration =	STRING_CDC,
 	.bmAttributes =		USB_CONFIG_ATT_ONE | USB_CONFIG_ATT_SELFPOWER,
-	.bMaxPower =		50,
+	.bMaxPower =		250,
 };
 
 #ifdef	CONFIG_USB_ETH_RNDIS
@@ -476,7 +476,7 @@
 	.bConfigurationValue =  DEV_RNDIS_CONFIG_VALUE,
 	.iConfiguration =       STRING_RNDIS,
 	.bmAttributes =		USB_CONFIG_ATT_ONE | USB_CONFIG_ATT_SELFPOWER,
-	.bMaxPower =            50,
+	.bMaxPower =            250,
 };
 #endif
 
