---
 sound/soc/s3c24xx/Kconfig |    2 	1 +	1 -	0 !
 1 file changed, 1 insertion(+), 1 deletion(-)

Index: linux-2.6/sound/soc/s3c24xx/Kconfig
===================================================================
--- linux-2.6.orig/sound/soc/s3c24xx/Kconfig	2007-07-23 10:15:13.000000000 +0200
+++ linux-2.6/sound/soc/s3c24xx/Kconfig	2007-07-23 10:18:07.000000000 +0200
@@ -18,7 +18,7 @@ config SND_S3C2443_SOC_AC97
 	
 config SND_S3C24XX_SOC_NEO1973_WM8753
 	tristate "SoC I2S Audio support for NEO1973 - WM8753"
-	depends on SND_S3C24XX_SOC && MACH_GTA01
+	depends on SND_S3C24XX_SOC && MACH_NEO1973_GTA01
 	select SND_S3C24XX_SOC_I2S
 	select SND_SOC_WM8753
 	help
