Index: linux-2.6.17/arch/arm/mach-versatile/core.c
===================================================================
--- linux-2.6.17.orig/arch/arm/mach-versatile/core.c
+++ linux-2.6.17/arch/arm/mach-versatile/core.c
@@ -396,8 +396,8 @@ static struct clcd_panel vga = {
 	.mode		= {
 		.name		= "VGA",
 		.refresh	= 60,
-		.xres		= 640,
-		.yres		= 480,
+		.xres		= 480,
+		.yres		= 640,
 		.pixclock	= 39721,
 		.left_margin	= 40,
 		.right_margin	= 24,
