diff -ibc matchbox-window-manager-1.1/src/structs.h matchbox-window-manager-1.1.patch/src/structs.h
*** matchbox-window-manager-1.1/src/structs.h	2006-07-25 16:16:05.000000000 -0400
--- matchbox-window-manager-1.1.patch/src/structs.h	2007-01-06 06:29:26.000000000 -0500
***************
*** 146,151 ****
--- 146,155 ----
  #define MB_CMD_COMPOSITE   8
  #define MB_CMB_KEYS_RELOAD 9
  
+ /* Defines for MB_COMMAND values, for openmoko project */
+ #define MB_CMD_ACTIVAE_CLIENT 100
+ #define MB_CMD_REMOVE_CLIENT 101
+ 
  /* Atoms, if you change these check ewmh_init() first */
  
  enum {
diff -ibc matchbox-window-manager-1.1/src/wm.c matchbox-window-manager-1.1.patch/src/wm.c
*** matchbox-window-manager-1.1/src/wm.c	2006-08-03 07:05:55.000000000 -0400
--- matchbox-window-manager-1.1.patch/src/wm.c	2007-01-06 06:28:15.000000000 -0500
***************
*** 1683,1688 ****
--- 1683,1699 ----
  	 case MB_CMD_DESKTOP   :
  	   wm_toggle_desktop(w);
  	   break;
+ /*For OpenMoko project, add by Sun Zhiyong*/	   
+ 	 case MB_CMD_ACTIVAE_CLIENT	:
+ 	   if (c == w->focused_client)
+ 	     break;
+ 	   else
+ 	     wm_activate_client(c);	   
+ 	     break;
+ 	 case MB_CMD_REMOVE_CLIENT	:
+ 	   wm_remove_client(w, c);
+ 	   break;
+ /*End ADD*/	   
  	 case MB_CMD_MISC:  /* This is used for random testing stuff */
  	   /* comp_engine_deinit(w); */
  #ifdef DEBUG
