This patch is a workaround of some S3C2410 MMC chip bug

Index: linux-2.6/drivers/mmc/host/s3cmci.c
===================================================================
--- linux-2.6.orig/drivers/mmc/host/s3cmci.c
+++ linux-2.6/drivers/mmc/host/s3cmci.c
@@ -445,9 +445,16 @@
 
 	if (mci_csta & S3C2410_SDICMDSTAT_CRCFAIL) {
 		if (cmd->flags & MMC_RSP_CRC) {
-			cmd->error = -EILSEQ;
-			host->status = "error: bad command crc";
-			goto fail_transfer;
+			if (host->mrq->cmd->flags & MMC_RSP_136) {
+				dbg(host, dbg_irq,
+				    "fixup: ignore CRC fail with long rsp\n");
+			} else {
+#if 0
+				cmd->error = -EILSEQ;
+				host->status = "error: bad command crc";
+				goto fail_transfer;
+#endif
+			}
 		}
 
 		mci_cclear |= S3C2410_SDICMDSTAT_CRCFAIL;
