Index: u-boot/cpu/arm920t/start.S
===================================================================
--- u-boot.orig/cpu/arm920t/start.S
+++ u-boot/cpu/arm920t/start.S
@@ -202,6 +202,13 @@
 #endif /* CONFIG_S3C2410_NAND_BOOT */
 
 relocate:				/* relocate U-Boot to RAM	    */
+#if !defined(CONFIG_SKIP_LOWLEVEL_INIT) && defined(CONFIG_LL_INIT_NAND_ONLY)
+	teq	r0, #0			/* running at address 0 ?	    */
+	bleq	cpu_init_crit		/* yes -> do low-level setup	    */
+#endif /* !CONFIG_SKIP_LOWLEVEL_INIT && CONFIG_LL_INIT_NAND_ONLY */
+
+	adr	r0, _start		/* the above may have clobbered r0  */
+
 	ldr	r1, _TEXT_BASE		/* test if we run from flash or RAM */
 	cmp     r0, r1                  /* don't reloc during debug         */
 	beq     done_relocate
