Index: u-boot/board/neo1973/gta02/gta02.c
===================================================================
--- u-boot.orig/board/neo1973/gta02/gta02.c
+++ u-boot/board/neo1973/gta02/gta02.c
@@ -306,12 +306,12 @@ continue_boot:
 		if (!(gpio->GPFDAT & (1 << 5)))
 			gpio->GPBDAT &= ~(1 << 2);
 	}
+#endif
 
 	if (menu_vote > 0) {
-		bootmenu();
+		neo1973_bootmenu();
 		nobootdelay = 1;
 	}
-#endif
 
 	return 0;
 }
@@ -408,13 +408,11 @@ int neo1973_on_key_pressed(void)
 						& PCF50633_OOCSTAT_ONKEY);
 }
 
-/* FIXME: shared */
 int neo1973_aux_key_pressed(void)
 {
 	S3C24X0_GPIO * const gpio = S3C24X0_GetBase_GPIO();
-	if (gpio->GPFDAT & (1 << 6))
-		return 0;
-	return 1;
+
+	return !!(gpio->GPFDAT & (1 << 6));
 }
 
 /* The sum of all part_size[]s must equal to or greater than the NAND size,
