Index: u-boot/cpu/arm920t/s3c24x0/mmc.c
===================================================================
--- u-boot.orig/cpu/arm920t/s3c24x0/mmc.c
+++ u-boot/cpu/arm920t/s3c24x0/mmc.c
@@ -60,8 +60,6 @@ static int wide = 0;
 static u_int32_t *mmc_cmd(ushort cmd, ulong arg, ushort flags)
 {
 	static u_int32_t resp[5];
-	ulong status;
-	int i;
 
 	u_int32_t ccon, csta;
 	u_int32_t csta_rdy_bit = S3C2410_SDICMDSTAT_CMDSENT;
@@ -134,7 +132,7 @@ static int mmc_block_read(uchar *dst, ul
 	//sdi->SDIPRE = 0xff;
 
 	/* setup data */
-	dcon = (len >> 9) & S3C2410_SDIDCON_BLKNUM_MASK;
+	dcon = (len >> 9) & S3C2410_SDIDCON_BLKNUM;
 	dcon |= S3C2410_SDIDCON_BLOCKMODE;
 	dcon |= S3C2410_SDIDCON_RXAFTERCMD|S3C2410_SDIDCON_XFER_RXSTART;
 	if (wide)
@@ -309,7 +307,7 @@ int mmc_write(uchar *src, ulong dst, int
 	return 0;
 }
 
-ulong mmc_bread(int dev_num, ulong blknr, ulong blkcnt, ulong *dst)
+ulong mmc_bread(int dev_num, ulong blknr, ulong blkcnt, void *dst)
 {
 	int mmc_block_size = MMC_BLOCK_SIZE;
 	ulong src = blknr * mmc_block_size + CFG_MMC_BASE;
@@ -318,7 +316,9 @@ ulong mmc_bread(int dev_num, ulong blknr
 	return blkcnt;
 }
 
-static u_int16_t rca = MMC_DEFAULT_RCA;
+/* MMC_DEFAULT_RCA should probably be just 1, but this may break other code
+   that expects it to be shifted. */
+static u_int16_t rca = MMC_DEFAULT_RCA >> 16;
 
 static u_int32_t mmc_size(const struct mmc_csd *csd)
 {
@@ -361,7 +361,6 @@ int mmc_init(int verbose)
 
 	printf("trying to detect SD Card...\n");
 	while (retries--) {
-		int i;
 		udelay(100000);
 		resp = mmc_cmd(55, 0x00000000, CMD_F_RESP);
 		resp = mmc_cmd(41, 0x00300000, CMD_F_RESP);
Index: u-boot/include/asm-arm/arch-s3c24x0/regs-sdi.h
===================================================================
--- u-boot.orig/include/asm-arm/arch-s3c24x0/regs-sdi.h
+++ u-boot/include/asm-arm/arch-s3c24x0/regs-sdi.h
@@ -62,7 +62,7 @@
 #define S3C2410_SDIDCON_XFER_RXSTART  (2<<12)
 #define S3C2410_SDIDCON_XFER_TXSTART  (3<<12)
 
-#define S3C2410_SDIDCON_BLKNUM_MASK   (0xFFF)
+#define S3C2410_SDIDCNT_BLKNUM_MASK   (0xFFF)
 #define S3C2410_SDIDCNT_BLKNUM_SHIFT  (12)
 
 #define S3C2410_SDIDSTA_RDYWAITREQ    (1<<10)
Index: u-boot/include/configs/neo1973.h
===================================================================
--- u-boot.orig/include/configs/neo1973.h
+++ u-boot/include/configs/neo1973.h
@@ -227,7 +227,7 @@
 #define CONFIG_EXT2		1
 
 #define CONFIG_FAT		1
-#define CONFIG_SUPPORT_VFAT	1
+#define CONFIG_SUPPORT_VFAT
 
 #if 1
 /* JFFS2 driver */
