This fixes a compile problem (no partition support for MMC)

Signed-off-by: Harald Welte <laforge@openmoko.org>

Index: u-boot.git/disk/part.c
===================================================================
--- u-boot.git.orig/disk/part.c	2007-01-13 17:59:35.000000000 +0100
+++ u-boot.git/disk/part.c	2007-01-13 17:59:45.000000000 +0100
@@ -126,6 +126,7 @@
 #if ((CONFIG_COMMANDS & CFG_CMD_IDE)	|| \
      (CONFIG_COMMANDS & CFG_CMD_SCSI)	|| \
      (CONFIG_COMMANDS & CFG_CMD_USB)	|| \
+     defined(CONFIG_MMC) || \
      defined(CONFIG_SYSTEMACE)          )
 
 #if defined(CONFIG_MAC_PARTITION) || \
