Index: u-boot.git/cpu/arm920t/s3c24x0/nand_read.c
===================================================================
--- u-boot.git.orig/cpu/arm920t/s3c24x0/nand_read.c	2007-01-18 23:49:55.000000000 +0100
+++ u-boot.git/cpu/arm920t/s3c24x0/nand_read.c	2007-01-19 00:04:21.000000000 +0100
@@ -33,6 +33,7 @@
 
 #define NAND_SECTOR_SIZE	512
 #define NAND_BLOCK_MASK		(NAND_SECTOR_SIZE - 1)
+#define NAND_PAGE_SIZE		0x4000
 
 /* low level nand read function */
 int
@@ -49,6 +50,26 @@
     for(i=0; i<10; i++);
 
     for(i=start_addr; i < (start_addr + size);) {
+
+#ifdef CONFIG_S3C2410_NAND_SKIP_BAD
+        if (start_addr % NAND_PAGE_SIZE == 0) {
+	    unsigned char data;
+	    NFCMD = 0x50;
+	    NFADDR = 517&0xf;
+	    NFADDR = (i >> 9) & 0xff;
+	    NFADDR = (i >> 17) & 0xff;
+	    NFADDR = (i >> 25) & 0xff;
+	    wait_idle();
+	    data = (NFDATA & 0xff);
+	    if (data != 0xff) {
+	    	/* Bad block */
+		i += NAND_PAGE_SIZE;
+		size += NAND_PAGE_SIZE;
+		continue;
+	    }
+	}
+#endif
+
       /* READ0 */
       NFCMD = 0;
 
Index: u-boot.git/include/configs/gta01.h
===================================================================
--- u-boot.git.orig/include/configs/gta01.h	2007-01-19 00:01:20.000000000 +0100
+++ u-boot.git/include/configs/gta01.h	2007-01-19 00:05:26.000000000 +0100
@@ -32,6 +32,7 @@
 #else
 /* we want to start u-boot directly from within NAND flash */
 #define CONFIG_S3C2410_NAND_BOOT	1
+#define CONFIG_S3C2410_NAND_SKIP_BAD	1
 #endif
 
 /*
