Index: u-boot/drivers/nand/nand_bbt.c
===================================================================
--- u-boot.orig/drivers/nand/nand_bbt.c
+++ u-boot/drivers/nand/nand_bbt.c
@@ -1054,7 +1054,13 @@ int nand_isbad_bbt (struct mtd_info *mtd
 }
 
 #define ARRAY_SIZE(x)           (sizeof(x) / sizeof((x)[0]))
-static unsigned int part_size[] = { 0x30000, 0x4000, 0x200000, 0x400000, 0x39cc000 };
+
+/* NOTE: this is extremely device-dependent, so it shouldn't really be here.
+   The sum of all part_size[]s must equal to the NAND size, i.e., 0x4000000.
+   "initrd" is sized such that it can hold two uncompressed 16 bit 640*480
+   images: 640*480*2*2 = 1228800 < 1245184. */
+
+static unsigned int part_size[] = { 0x30000, 0x4000, 0x200000, 0x130000, 0x3c9c000 };
 static char *part_names[] = { "u-boot", "u-boot env", "kernel", "initrd", "rootfs" };
 
 int nand_create_mtd_dynpart(struct mtd_info *mtd)
