Index: u-boot/cpu/arm920t/s3c24x0/nand.c
===================================================================
--- u-boot.orig/cpu/arm920t/s3c24x0/nand.c
+++ u-boot/cpu/arm920t/s3c24x0/nand.c
@@ -122,7 +122,9 @@
 }
 #endif
 
-int board_nand_init(struct nand_chip *nand)
+int board_nand_init(void) __attribute__((weak, alias("__board_nand_init")));
+
+int __board_nand_init(struct nand_chip *nand)
 {
 	u_int32_t cfg;
 	u_int8_t tacls, twrph0, twrph1;
Index: u-boot/include/s3c24x0.h
===================================================================
--- u-boot.orig/include/s3c24x0.h
+++ u-boot/include/s3c24x0.h
@@ -1122,4 +1122,6 @@
 
 #endif
 
+int __board_nand_init(struct nand_chip *nand);
+
 #endif /*__S3C24X0_H__*/
