Fix building with CRAMFS but not JFFS2 support

Signed-off-by: Harald Welte <laforge@openmoko.org>

Index: u-boot.git/fs/cramfs/cramfs.c
===================================================================
--- u-boot.git.orig/fs/cramfs/cramfs.c	2007-02-07 16:47:39.000000000 +0100
+++ u-boot.git/fs/cramfs/cramfs.c	2007-02-07 16:48:01.000000000 +0100
@@ -27,7 +27,7 @@
 #include <common.h>
 #include <malloc.h>
 
-#if (CONFIG_COMMANDS & CFG_CMD_JFFS2)
+#if (CONFIG_COMMANDS & CFG_CMD_CRAMFS)
 
 #include <asm/byteorder.h>
 #include <linux/stat.h>
@@ -343,5 +343,22 @@
 	}
 	return 1;
 }
-
+#else
+int cramfs_load (char *loadoffset, struct part_info *info, char *filename)
+{
+	return -1;
+}
+int cramfs_ls (struct part_info *info, char *filename)
+{
+	return 0;
+}
+int cramfs_info (struct part_info *info)
+{
+	return 0;
+}
+int cramfs_check (struct part_info *info)
+{
+	return 0;
+}
 #endif /* CFG_FS_CRAMFS */
+

