Resume support for low-level uboot code, Version 5

Signed-off-by: Ben Dooks <ben-linux@fluff.org>

Index: u-boot/board/qt2410/lowlevel_init.S
===================================================================
--- u-boot.orig/board/qt2410/lowlevel_init.S	2007-02-21 10:58:40.000000000 +0100
+++ u-boot/board/qt2410/lowlevel_init.S	2007-02-21 15:54:14.000000000 +0100
@@ -42,6 +42,8 @@
  *
  */
 
+#include "s3c2410.h"
+	
 #define BWSCON	0x48000000
 
 /* BWSCON */
@@ -132,6 +134,17 @@
 
 .globl lowlevel_init
 lowlevel_init:
+	/* take sdram out of power down */
+	ldr	r0, =0x56000080		/* misccr */
+	ldr	r1, [ r0 ]
+	bic	r1, r1, #(S3C2410_MISCCR_nEN_SCLK0 | S3C2410_MISCCR_nEN_SCLK1 | S3C2410_MISCCR_nEN_SCLKE)
+	str	r1, [ r0 ]
+
+	/* ensure signals stabalise */
+	mov	r1, #128
+1:	subs	r1, r1, #1
+	bpl	1b
+	
 	/* memory control configuration */
 	/* make r0 relative the current location so that it */
 	/* reads SMRDATA out of FLASH rather than memory ! */
@@ -150,6 +163,21 @@
 	mrc	p15, 0, r1 ,c1 ,c0, 0
 	orr	r1, r1, #0xc0000000
 	mcr	p15, 0, r1, c1, c0, 0
+	
+	/* ensure some refresh has happened */
+	mov	r1, #4096
+1:	subs	r1, r1, #1
+	bpl	1b
+	
+	/* test for resume */
+	ldr	r1, =0x560000B4		/* gstatus2 */
+	ldr	r0, [ r1 ]
+	tst	r0, #0x02		/* is this resume from power down */
+	beq	not_resuming
+
+	ldr	pc, [r1, #4]		/* gstatus3 */
+	
+not_resuming:
 
 	/* everything is fine now */
 	mov	pc, lr
