Index: u-boot/board/neo1973/gta02/config.mk
===================================================================
--- u-boot.orig/board/neo1973/gta02/config.mk
+++ u-boot/board/neo1973/gta02/config.mk
@@ -24,7 +24,7 @@
 #
 # download area is 3200'0000 or 3300'0000
 
-CONFIG_USB_DFU_VENDOR=0x1457
+CONFIG_USB_DFU_VENDOR=0x1d50
 CONFIG_USB_DFU_PRODUCT=0x5119
 
 sinclude $(OBJTREE)/board/$(BOARDDIR)/config.tmp
Index: u-boot/include/configs/neo1973_gta02.h
===================================================================
--- u-boot.orig/include/configs/neo1973_gta02.h
+++ u-boot/include/configs/neo1973_gta02.h
@@ -177,7 +177,7 @@
 #define CONFIG_USB_DEVICE	1
 #define CONFIG_USB_TTY		1
 #define CFG_CONSOLE_IS_IN_ENV	1
-#define CONFIG_USBD_VENDORID 		0x1457     /* Linux/NetChip */
+#define CONFIG_USBD_VENDORID 		0x1d50    /* OpenMoko, Inc. */
 #define CONFIG_USBD_PRODUCTID_GSERIAL	0x5120    /* gserial */
 #define CONFIG_USBD_PRODUCTID_CDCACM 	0x5119    /* CDC ACM */
 #define CONFIG_USBD_MANUFACTURER	"OpenMoko, Inc"
